examples/rag/use_own_knowledge_dataset.py

```diff
@@ -0,0 +1,125 @@
+    """Compute the DPR embeddings of document passages"""
+    input_ids = ctx_tokenizer(
+        documents["title"], documents["text"], truncation=True, padding="longest", return_tensors="pt"
+    )["input_ids"]
+    embeddings = ctx_encoder(input_ids.to(device=device), return_dict=True).pooler_output
+    return {"embeddings": embeddings.detach().cpu().numpy()}
+
+
+def main(
+    tmp_dir: str,
+    num_proc=2,
+    batch_size=16,
+    rag_model_name="facebook/rag-sequence-nq",
+    dpr_ctx_encoder_model_name="facebook/dpr-ctx_encoder-single-nq-base",
+):
+
+    ######################################
+    print("Step 1 - Create the dataset")
+    ######################################
+
+    # The dataset needed for RAG must have three columns:
+    # - title (string): title of the document
+    # - text (string): text of a passage of the document
+    # - embeddings (array of dimension d): DPR representation of the passage
+
+    # Let's say you have documents in tab-separated csv files with columns "title" and "text"
+    csv_path = str(Path(__file__).parent / "test_data" / "my_knowledge_dataset.csv")
+
+    # You can load a Dataset object this way
+    dataset = load_dataset("csv", data_files=[csv_path], split="train", delimiter="\t", column_names=["title", "text"])
+
+    # Then split the documents into passages of 100 words
+    dataset = dataset.map(split_documents, batched=True, num_proc=num_proc)
+
+    # And compute the embeddings
+    # TODO(QL): Use the DPR context encoder trained on the multiset/hybrid dataset instead of single NQ
+    ctx_encoder = DPRContextEncoder.from_pretrained(dpr_ctx_encoder_model_name).to(device=device)
+    ctx_tokenizer = DPRContextEncoderTokenizerFast.from_pretrained(dpr_ctx_encoder_model_name)
+    dataset = dataset.map(
+        partial(embed, ctx_encoder=ctx_encoder, ctx_tokenizer=ctx_tokenizer), batched=True, batch_size=batch_size
+    )
+
+    # And finally save your dataset
+    dataset_path = os.path.join(tmp_dir, "my_knowledge_dataset")
+    dataset.save_to_disk(dataset_path)
+
+    ######################################
+    print("Step 2 - Index the dataset")
+    ######################################
+
+    # Let's use the Faiss implementation of HNSW for fast approximate nearest neighbor search
+    d = 768  # vectors dimension
+    m = 128  # hnsw parameter. Higher is more accurate but takes more time to index (default is 32, 128 should be ok)
+    index = faiss.IndexHNSWFlat(d, m, faiss.METRIC_INNER_PRODUCT)
+    dataset.add_faiss_index("embeddings", custom_index=index)
+
+    # And save the index
+    index_path = os.path.join(tmp_dir, "my_knowledge_dataset_hnsw_index.faiss")
+    dataset.get_index("embeddings").save(index_path)
+
+    ######################################
+    print("Step 3 - Load RAG")
+    ######################################
+
+    retriever = RagRetriever.from_pretrained(
+        rag_model_name, index_name="custom", dataset=dataset_path, index_path=index_path
```

Comment:
Could we directly load a dataset with an index and prevent the user from having to save and load the index separately?

Reply:
Sure we can. In that case we can just save the dataset and the faiss index when calling `save_pretrained`.
I'll see what I can do

Reply:
I added a way to pass the indexed dataset directly, and I made it compatible with save/load_pretrained